identity/src/peer_id.rs

```diff
@@ -79,6 +79,21 @@ impl PeerId {
         PeerId { multihash }
     }
 
+    /// Extracts the `PublicKey` from a `PeerId`.
+    /// Only works for Identity Multihash PeerIds from ed25519 keys
+    /// Errors if the PeerId is not an Identity Multihash or decoding fails
+    pub fn to_public_key(&self) -> Result<crate::keypair::PublicKey, Box<dyn std::error::Error>> {
+        if self.multihash.code() != MULTIHASH_IDENTITY_CODE {
+            Err(Box::new(Error::UnsupportedCode(self.multihash.code())))
+        } else {
+            let key_enc = self.multihash.digest();
+            match crate::keypair::PublicKey::try_decode_protobuf(key_enc) {
+                Ok(key) => Ok(key),
+                Err(err) => Err(Box::new(err)),
+            }
+        }
+    }
```

Comment:
I can see the appeal of providing this function here but I am not quite happy with the API and implementation.

- I don't want to expose raw error trait objects. It is too easy to get the auto-traits wrong on this. Can we reuse our `DecodingError`?
- The naming needs to follow conventions (i.e. `try_to` for fns that return `Result`)
- I'd prefer early return instead of `else`
- Rustdoc should follow convention of a one-line heading with a newline before the paragraph starts.

If you feel strongly about including this, please address the above :)
Otherwise happy to fastpath the other changes into master and we discuss this in a separate PR.

Reply:
All good points, it was indeed a bit of a challenge to quickly squeeze this into this PR. I am happy to address this in a separate PR to do it right.

The problem was it gives potentially 2 error codes 1) multihash UnsupportedCode an 2) DecodingError which we have to cover in the Result (...in a separate PR)